packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts

```diff
@@ -0,0 +1,236 @@
+      .flatMap((requirement) => requirement.submission)
+      .forEach((submission, index) => {
+        const verifiableCredential = submission.verifiableCredentials[
+          options.submissionEntryIndexes[index] as number
+        ] as W3cCredentialRecord
+
+        const inputDescriptor = credentialsForInputDescriptor[submission.inputDescriptorId]
+        if (!inputDescriptor) {
+          credentialsForInputDescriptor[submission.inputDescriptorId] = [verifiableCredential.credential]
+        } else {
+          inputDescriptor.push(verifiableCredential.credential)
+        }
+      })
+
+    const vps = await this.presentationExchangeService.createPresentation(agentContext, {
+      credentialsForInputDescriptor,
+      presentationDefinition: options.verifiedAuthorizationRequest.presentationDefinitions[0].definition,
+      includePresentationSubmissionInVp: false,
+      // TODO: are there other properties we need to include?
+      nonce: await options.verifiedAuthorizationRequest.authorizationRequest.getMergedProperty<string>('nonce'),
+    })
+
+    const verificationMethod = await this.getVerificationMethodFromVerifiablePresentation(
+      agentContext,
+      vps.verifiablePresentations[0] as W3cVerifiablePresentation
+    )
+    const key = getKeyFromVerificationMethod(verificationMethod)
+    const alg = getJwkClassFromKeyType(key.keyType)?.supportedSignatureAlgorithms[0]
+    if (!alg) {
+      throw new AriesFrameworkError(`No supported algs for key type: ${key.keyType}`)
+    }
+
+    const response = await op.createAuthorizationResponse(options.verifiedAuthorizationRequest, {
+      issuer: verificationMethod.controller,
+      presentationExchange: {
+        verifiablePresentations: vps.verifiablePresentations.map((vp) => vp.encoded as W3CVerifiablePresentation),
+        presentationSubmission: vps.presentationSubmission,
+      },
+      signature: {
+        signature: async (data) => {
+          const signature = await agentContext.wallet.sign({
+            data: typeof data === 'string' ? TypedArrayEncoder.fromString(data) : Buffer.from(data),
+            key,
+          })
+
+          return TypedArrayEncoder.toBase64URL(signature)
+        },
+        // FIXME: cast
+        alg: alg as unknown as SigningAlgo,
+        did: verificationMethod.controller,
+        kid: verificationMethod.id,
+      },
+    })
+
+    const responseToResponse = await op.submitAuthorizationResponse(response)
+
+    if (!responseToResponse.ok) {
+      throw new AriesFrameworkError(`Error submitting authorization response. ${await responseToResponse.text()}`)
+    }
+  }
+
+  private getSupportedDidMethods(agentContext: AgentContext) {
+    const didsApi = agentContext.dependencyManager.resolve(DidsApi)
+    const supportedDidMethods: string[] = []
+
+    for (const resolver of didsApi.config.resolvers) {
+      supportedDidMethods.push(...resolver.supportedMethods)
+    }
+
+    return supportedDidMethods
```

Comment:
Might be good to filter out duplicates here. Very minor comment though.

Reply:
I am using a set instead of an array now